### ISSUE TITLE
Statically linked rust executable segfaults

### ISSUE THREAD

**Issue by TerrorJack** [chimera-linux/cports#2511]
In the `chimeralinux/chimera` container:

```sh
apk add cargo libcxx-devel-static
RUSTFLAGS="-C target-feature=+crt-static" cargo install --target x86_64-chimera-linux-musl --root $PWD names
./bin/names 
```

This would result in `Segmentation fault (core dumped)`. Also I noticed that `musl-devel-static` wasn't needed when linking the final executable, so it seems the rust toolchain packaged in chimera indeed uses its own libc archive when static linking instead of system-wide `libc.a`, and that archive might be problematic.

**Comment by nekopsykose** [comment]
> so it seems the rust toolchain packaged in chimera indeed uses its own libc archive when static linking instead of system-wide libc.a

if you look at the link invocation what most likely happens here is it doesn't link libc.a at all

**Comment by nekopsykose** [comment]
these segfaulting programs are when the output is not actually static, you can readelf -d it or run `file` on it and it's still dynamically linked, but the linker invocation passes the static runtime (that's what +crt-static stands for, among other things). and having a dynamic exe with the static runtime segfaults if the program has dso deps (the static runtime does not read DT_NEEDED and load dsos), so then the external symbols are null and it segfaults trying to call a library function

.. is the case 99% of the time. and it's also an upstream rust issue, but i thought they at least fixed outputting a broken static executable. if you want me to look at it you'll have to post what `names` is and all the code to reproduce this with

**Comment by TerrorJack** [comment]
Thanks for the rapid answer! Yes, I can confirm with `file` that the output is still dynamically linked, and it might be an upstream issue, though this used to work in alpine so I thought it's the same in chimera as well.

A much simpler repro is:

```rust
fn main() {
    println!("Hello, world!");
}
```

And compiling it with `rustc hello.rs -C target-feature=+crt-static`.

**Comment by nekopsykose** [comment]
looking at the link invocation:  

```
/usr/bin/cc
-m64
/tmp/rustcP1uI2S/symbols.o
main.main.a2fbdd81ee48c0a9-cgu.0.rcgu.o
main.tlpf16i1edh2wz9.rcgu.o
-Wl,--as-needed
-L
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib
-Wl,-Bstatic
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libstd-c8a6cac7cd5963a3.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libpanic_unwind-f82afa78c15b9f0c.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libobject-87d091b5e67ce742.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libmemchr-1fc91eff9710663d.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libaddr2line-5f4685ca65ede4dd.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libgimli-bf6e3879d32ef033.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/librustc_demangle-442ea6c6292c5221.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libstd_detect-6093e36c57be79b8.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libhashbrown-2057f70b8f105c25.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/librustc_std_workspace_alloc-23d8a7c5054c464e.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libminiz_oxide-39d27e4b83ef7bb9.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libadler-2b8f426efe9f8a5c.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libunwind-f6cef70a49889af6.rlib
-lunwind
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libcfg_if-5f71d504767904e9.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/liblibc-387841a5e274f573.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/liballoc-0539b405f9fe2309.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/librustc_std_workspace_core-19e04ebde3687a0e.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libcore-f9d94f527459ad0b.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libcompiler_builtins-22101762fe7fe286.rlib
-Wl,-Bdynamic
-Wl,--eh-frame-hdr
-Wl,-z,noexecstack
-L
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib
-o
main
-Wl,--gc-sections
-static-pie
-Wl,-z,relro,-z,now
'-fuse-ld=lld'

/usr/bin/ld.lld
-z
now
-z
relro
'--hash-style=gnu'
--build-id
--eh-frame-hdr
-m
elf_x86_64
-static
-pie
--no-dynamic-linker
-z
text
-o
main
/lib/rcrt1.o
/lib/crti.o
/usr/lib/clang/18/lib/x86_64-chimera-linux-musl/clang_rt.crtbegin.o
-L/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib
-L/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib
-L/usr/lib/clang/18/lib/x86_64-chimera-linux-musl
-L/lib/
-L/lib/../lib64
-L/usr/lib/
-L/usr/lib/../lib64
-L/lib
-L/usr/lib
--as-needed
/tmp/rustcP1uI2S/symbols.o
main.main.a2fbdd81ee48c0a9-cgu.0.rcgu.o
main.tlpf16i1edh2wz9.rcgu.o
--as-needed
-Bstatic
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libstd-c8a6cac7cd5963a3.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libpanic_unwind-f82afa78c15b9f0c.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libobject-87d091b5e67ce742.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libmemchr-1fc91eff9710663d.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libaddr2line-5f4685ca65ede4dd.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libgimli-bf6e3879d32ef033.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/librustc_demangle-442ea6c6292c5221.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libstd_detect-6093e36c57be79b8.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libhashbrown-2057f70b8f105c25.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/librustc_std_workspace_alloc-23d8a7c5054c464e.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libminiz_oxide-39d27e4b83ef7bb9.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libadler-2b8f426efe9f8a5c.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libunwind-f6cef70a49889af6.rlib
-lunwind
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libcfg_if-5f71d504767904e9.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/liblibc-387841a5e274f573.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/liballoc-0539b405f9fe2309.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/librustc_std_workspace_core-19e04ebde3687a0e.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libcore-f9d94f527459ad0b.rlib
/usr/lib64/rustlib/x86_64-chimera-linux-musl/lib/libcompiler_builtins-22101762fe7fe286.rlib
-Bdynamic
--eh-frame-hdr
-z
noexecstack
--gc-sections
-z
relro
-z
now
--start-group
/usr/lib/clang/18/lib/x86_64-chimera-linux-musl/libclang_rt.builtins.a
-l:libunwind.a
--push-state
--as-needed
-latomic
--pop-state
-lc
--end-group
/usr/lib/clang/18/lib/x86_64-chimera-linux-musl/clang_rt.crtend.o
/lib/crtn.o
```

on alpine:  

```
cc
-m64
/tmp/rustcaNYGKD/symbols.o
main.main.172bdf8e90170551-cgu.0.rcgu.o
main.1uktdjnqi0fsrc75.rcgu.o
-Wl,--as-needed
-L
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib
-Wl,-Bstatic
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libstd-54bb4f6d7fc61a8d.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libpanic_unwind-d7ee381ff2952b1b.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libobject-717898023ba5ce45.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libmemchr-216051883822340c.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libaddr2line-1a87e02d85b0398f.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libgimli-4f3983cad535c23c.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/librustc_demangle-a7e690a7cf68cbf6.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libstd_detect-ecc44c11554e7cb5.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libhashbrown-6f65b4e0ce2a6c7e.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/librustc_std_workspace_alloc-bde77491879bcee4.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libminiz_oxide-523c092783f39e0a.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libadler-3484acf5667268cf.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libunwind-657e06d6f5138488.rlib
-lgcc_eh
-lc
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libcfg_if-b1784829a60e1fbb.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/liblibc-e70ff63fef49d077.rlib
-lc
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/liballoc-d2d2d58fa83a60d6.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/librustc_std_workspace_core-c505507c02a386f6.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libcore-394a3a469db7e251.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libcompiler_builtins-f1ee18e2cb9913d5.rlib
-Wl,-Bdynamic
-Wl,--eh-frame-hdr
-Wl,-z,noexecstack
-L
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib
-o
main
-Wl,--gc-sections
-static-pie
-Wl,-z,relro,-z,now
-nodefaultlibs
'-fuse-ld=lld'

/usr/bin/ld.lld
-plugin
/usr/libexec/gcc/x86_64-alpine-linux-musl/13.2.1/liblto_plugin.so
'-plugin-opt=/usr/libexec/gcc/x86_64-alpine-linux-musl/13.2.1/lto-wrapper'
'-plugin-opt=-fresolution=/tmp/ccoAcjAj.res'
--build-id
--eh-frame-hdr
'--hash-style=gnu'
-m
elf_x86_64
--as-needed
-static
-pie
--no-dynamic-linker
-z
text
-pie
-z
now
-o
main
/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../lib/rcrt1.o
/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../lib/crti.o
/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/crtbeginS.o
-L/usr/lib/rustlib/x86_64-alpine-linux-musl/lib
-L/usr/lib/rustlib/x86_64-alpine-linux-musl/lib
-L/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1
-L/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../x86_64-alpine-linux-musl/lib/../lib
-L/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../lib
-L/lib/../lib
-L/usr/lib/../lib
-L/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../x86_64-alpine-linux-musl/lib
-L/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../..
/tmp/rustcaNYGKD/symbols.o
main.main.172bdf8e90170551-cgu.0.rcgu.o
main.1uktdjnqi0fsrc75.rcgu.o
--as-needed
-Bstatic
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libstd-54bb4f6d7fc61a8d.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libpanic_unwind-d7ee381ff2952b1b.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libobject-717898023ba5ce45.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libmemchr-216051883822340c.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libaddr2line-1a87e02d85b0398f.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libgimli-4f3983cad535c23c.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/librustc_demangle-a7e690a7cf68cbf6.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libstd_detect-ecc44c11554e7cb5.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libhashbrown-6f65b4e0ce2a6c7e.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/librustc_std_workspace_alloc-bde77491879bcee4.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libminiz_oxide-523c092783f39e0a.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libadler-3484acf5667268cf.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libunwind-657e06d6f5138488.rlib
-lgcc_eh
-lc
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libcfg_if-b1784829a60e1fbb.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/liblibc-e70ff63fef49d077.rlib
-lc
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/liballoc-d2d2d58fa83a60d6.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/librustc_std_workspace_core-c505507c02a386f6.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libcore-394a3a469db7e251.rlib
/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libcompiler_builtins-f1ee18e2cb9913d5.rlib
-Bdynamic
--eh-frame-hdr
-z
noexecstack
--gc-sections
-z
relro
-z
now
/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/crtendS.o
/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../lib/crtn.o
```

so, rust does some Weird Stuff wrt link args where they mix -Bdynamic/-Bstatic a bunch and when you do +crt-static they still keep using -Bdynamic (but in different ordering), and so the implicit `-lc` link ends up being dynamic in our case. normally, they inject `-lc` themselves in the right spot..

..so the reason this ends up breaking is purely because of a downstream patch we have (i.e. this works on alpine with -Clinker=clang or whatever) in 0005-get-rid-of-wrong-linkage.patch.  it removes rusts entire `-lc` injection (which would come right after `-Wl,-Bstatic` and before `-Wl,-Bdynamic`) which happens to link in libc statically and make a working exe. it never ends up linking it at all in our case, using the implicit -lc which would be dynamic because of the other args

so yeah, this is a downstream patch bug and all static linkage is broken with the rustc here. but the reason that patch exists is because without it rustc will try to link in builtins for you, so it expects for e.g. gcc_s to exist and it doesn't handle the clang layout with just libunwind/clang-builtins(from compiler-rt) (i.e. by default it does not work on a pure-clang setup instead of one where libgcc_s still exists and whatnot like on most gcc-based distros, or at least that's the case without the patch).  

the tl;dr i guess is that this isn't fixable, unless somehow rustc starts working by default on a pure-clang/llvm setup so the patch can be removed, which is unfortunate

..it would also be nice if the rustc link invoker did not juggle Bdynamic/Bstatic like this, which would also work, since if there was no -Bdynamic on the linker invocation at all (just remove it and try, you can find out the invocation via `extrace` or some obscure rustc debug arg i don't remember) this entire thing would happen to work anyway, since nothing would be confusing the linker into making a dynamic exe with it (with just -static and friends it will happily give you a static exe)
